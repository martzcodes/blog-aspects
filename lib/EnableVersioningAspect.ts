import { IAspect } from "aws-cdk-lib";
import { CfnBucket } from "aws-cdk-lib/aws-s3";
import { IConstruct } from "constructs";

export class EnableVersioningAspect implements IAspect {
  public visit(node: IConstruct): void {
    if (node instanceof CfnBucket && !node.versioningConfiguration) {
      node.versioningConfiguration = {
        status: 'Enabled'
      };
    }
  }
}