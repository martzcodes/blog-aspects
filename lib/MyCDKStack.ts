import { Stack, StackProps } from 'aws-cdk-lib';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

export class MyCDKStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    new Bucket(this, 'MyBucket', {
      versioned: false,
    });
  }
}