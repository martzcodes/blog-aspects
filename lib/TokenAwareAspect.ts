import { IAspect } from "aws-cdk-lib";
import { Bucket } from "aws-cdk-lib/aws-s3";
import { IConstruct } from "constructs";

export class TokenAwareAspect implements IAspect {
  visit(node: IConstruct): void {
    if (node instanceof Bucket) {
      console.log(`Bucket name is ${node.bucketName}`);
    }
  }
}