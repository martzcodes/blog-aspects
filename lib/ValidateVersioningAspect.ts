import { IAspect } from "aws-cdk-lib";
import { CfnBucket } from "aws-cdk-lib/aws-s3";
import { IConstruct } from "constructs";

export class ValidateVersioningAspect implements IAspect {
  public visit(node: IConstruct): void {
    if (node instanceof CfnBucket && !node.versioningConfiguration) {
      throw new Error(`Bucket ${node.node.path} does not have versioning enabled`);
    }
  }
}