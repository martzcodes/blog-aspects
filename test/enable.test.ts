import { App, Aspects } from "aws-cdk-lib";
import { MyCDKStack } from "../lib/MyCDKStack";
import { EnableVersioningAspect } from "../lib/EnableVersioningAspect";
import { Match, Template } from "aws-cdk-lib/assertions";

test("Bucket does not have versioning without aspect", () => {
  const app = new App();
  const stack = new MyCDKStack(app, "MyTestStack");
  const template = Template.fromStack(stack);

  template.resourceCountIs("AWS::S3::Bucket", 1);
  template.resourcePropertiesCountIs(
    "AWS::S3::Bucket",
    Match.serializedJson({
      VersioningConfiguration: Match.absent(),
    }),
    0
  );
  expect(template).toMatchSnapshot();
});

test("Bucket does have versioning with aspect", () => {
  const app = new App();
  const stack = new MyCDKStack(app, "MyTestStack");
  Aspects.of(stack).add(new EnableVersioningAspect());
  const template = Template.fromStack(stack);

  template.resourceCountIs("AWS::S3::Bucket", 1);
  template.resourcePropertiesCountIs(
    "AWS::S3::Bucket",
    {
      VersioningConfiguration: { Status: "Enabled" },
    },
    1
  );
  expect(template).toMatchSnapshot();
});
