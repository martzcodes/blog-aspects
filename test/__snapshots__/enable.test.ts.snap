// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bucket does have versioning with aspect 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "MyBucketF68F3FF0": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "VersioningConfiguration": {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`Bucket does not have versioning without aspect 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "MyBucketF68F3FF0": {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
